### ISSUE TITLE
Invalid git fact execution across multiple hosts

### ISSUE THREAD

**Issue by blarghmatey** [Fizzadar/pyinfra#560]
## Describe the bug
When using the `git.repo` operation and passing a destination path that relies on `host.data` attributes for the same repository across 2 hosts. After the initial run, when I try to execute the full deploy again it errors due to using the git_branch fact [here](https://github.com/blarghmatey/pyinfra/blob/master/pyinfra/operations/git.py#L125) because it's saying that it can't find the directory specified for the other host. 

## To Reproduce
The minimal code to reproduce the error is:
`git_test.py`
```
from pyinfra import host
from pyinfra.operations import git

git.repo(
    name="Clone Wordpress site repository",
    src="https://gitlab.com/BoundlessNotions/podcast-wp",
    dest=f"/var/www/{host.data.podcast}",
    branch="master",
    user="wordpress",
)
```
`inventory.py`
```
wordpress_servers = [
    ("66.175.213.219", {"podcast": "podcastinit"}),
    ("172.104.214.53", {"podcast": "dataengineeringpodcast"})
]
```

## Expected behavior
The expectation is that the deployment will clone the repository to the proper directory on each host without erroring on subsequent runs.

## Meta
+ Include output of `pyinfra --support`.
```
--> Support information:
    System: Linux

    If you are having issues with pyinfra or wish to make feature requests, please
    check out the GitHub issues at https://github.com/Fizzadar/pyinfra/issues .
    When adding an issue, be sure to include the following:

      Platform: Linux-5.11.8-1-MANJARO-x86_64-with-glibc2.33
      Release: 5.11.8-1-MANJARO
      Machine: x86_64
    pyinfra: v1.3.9
    Executable: /home/tmacey/code/boundlessnotions/config-management/.venv/bin/pyinfra
    Python: 3.9.2 (CPython, GCC 10.2.0)
```
+ How was pyinfra installed (source/pip)? via pip
+ Include pyinfra-debug.log (if one was created)
+ Consider including output with `-vv` and `--debug`.
```
pyinfra inventory.py git_test.py --debug
    [pyinfra_cli.main] Checking potential directory: src/bn_config/deployments
    [pyinfra_cli.main] Checking potential directory: src/bn_config
    [pyinfra_cli.main] Checking potential directory: src/bn_config/deployments
    [pyinfra_cli.main] Checking potential directory: src/bn_config
--> Loading config...
--> Loading inventory...
    [pyinfra_cli.inventory] Creating fake inventory...

--> Connecting to hosts...
    [pyinfra.api.connectors.ssh] Connecting to: 66.175.213.219 ({'allow_agent': True, 'look_for_keys': True, 'hostname': '66.175.213.219', 'username': 'root', 'timeout': 10})
    [pyinfra.api.connectors.ssh] Connecting to: 172.104.214.53 ({'allow_agent': True, 'look_for_keys': True, 'hostname': '172.104.214.53', 'username': 'root', 'timeout': 10})
    [172.104.214.53] Connected
    [66.175.213.219] Connected
    [pyinfra.api.state] Activating host: 66.175.213.219
    [pyinfra.api.state] Activating host: 172.104.214.53

--> Preparing operations...
    Loading: src/bn_config/deployments/git_test.py
    [pyinfra.api.operation] Adding operation, {'Clone Wordpress site repository'}, opOrder=(4,), opHash=6a58b6c7e02f6d92150e84bffa4418d987f54dc9
    [pyinfra.api.facts] Getting fact: directory (path=/var/www/podcastinit) (ensure_hosts: (Host(66.175.213.219),))
    [pyinfra.api.connectors.ssh] Running command on 66.175.213.219: (pty=False) sh -c '! test -e /var/www/podcastinit || (stat -c '"'"'user=%U group=%G mode=%A atime=%X mtime=%Y ctime=%Z size=%s %N'"'"' /var/www/podcastinit 2> /dev/null || stat -f '"'"'user=%Su group=%Sg mode=%Sp atime=%a mtime=%m ctime=%c size=%z %N%SY'"'"' /var/www/podcastinit)'
    [pyinfra.api.connectors.ssh] Running command on 172.104.214.53: (pty=False) sh -c '! test -e /var/www/podcastinit || (stat -c '"'"'user=%U group=%G mode=%A atime=%X mtime=%Y ctime=%Z size=%s %N'"'"' /var/www/podcastinit 2> /dev/null || stat -f '"'"'user=%Su group=%Sg mode=%Sp atime=%a mtime=%m ctime=%c size=%z %N%SY'"'"' /var/www/podcastinit)'
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 0
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 0
    [pyinfra.api.facts] Loaded fact directory (path=/var/www/podcastinit)
    [pyinfra.api.host] [66.175.213.219] noop: directory /var/www/podcastinit already exists
    [pyinfra.api.facts] Getting fact: directory (path=/var/www/podcastinit/.git) (ensure_hosts: (Host(66.175.213.219),))
    [pyinfra.api.connectors.ssh] Running command on 66.175.213.219: (pty=False) sh -c '! test -e /var/www/podcastinit/.git || (stat -c '"'"'user=%U group=%G mode=%A atime=%X mtime=%Y ctime=%Z size=%s %N'"'"' /var/www/podcastinit/.git 2> /dev/null || stat -f '"'"'user=%Su group=%Sg mode=%Sp atime=%a mtime=%m ctime=%c size=%z %N%SY'"'"' /var/www/podcastinit/.git)'
    [pyinfra.api.connectors.ssh] Running command on 172.104.214.53: (pty=False) sh -c '! test -e /var/www/podcastinit/.git || (stat -c '"'"'user=%U group=%G mode=%A atime=%X mtime=%Y ctime=%Z size=%s %N'"'"' /var/www/podcastinit/.git 2> /dev/null || stat -f '"'"'user=%Su group=%Sg mode=%Sp atime=%a mtime=%m ctime=%c size=%z %N%SY'"'"' /var/www/podcastinit/.git)'
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 0
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 0
    [pyinfra.api.facts] Loaded fact directory (path=/var/www/podcastinit/.git)
    [pyinfra.api.facts] Getting fact: git_branch (repo=/var/www/podcastinit) (ensure_hosts: (Host(66.175.213.219),))
    [pyinfra.api.connectors.ssh] Running command on 66.175.213.219: (pty=False) sh -c '! command -v git >/dev/null || cd /var/www/podcastinit && git rev-parse --abbrev-ref HEAD'
    [pyinfra.api.connectors.ssh] Running command on 172.104.214.53: (pty=False) sh -c '! command -v git >/dev/null || cd /var/www/podcastinit && git rev-parse --abbrev-ref HEAD'
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 2
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 0
    [172.104.214.53] sh: 1: cd: can't cd to /var/www/podcastinit
    [172.104.214.53] Error: could not load fact: git_branch (repo=/var/www/podcastinit)
    [pyinfra.api.facts] Loaded fact git_branch (repo=/var/www/podcastinit)
    [pyinfra.api.state] Failing hosts: 172.104.214.53
    [66.175.213.219] Ready: src/bn_config/deployments/git_test.py
    [pyinfra.api.operation] Adding operation, {'Clone Wordpress site repository'}, opOrder=(4,), opHash=6a58b6c7e02f6d92150e84bffa4418d987f54dc9
    [pyinfra.api.facts] Getting fact: directory (path=/var/www/dataengineeringpodcast) (ensure_hosts: (Host(172.104.214.53),))
    [pyinfra.api.connectors.ssh] Running command on 66.175.213.219: (pty=False) sh -c '! test -e /var/www/dataengineeringpodcast || (stat -c '"'"'user=%U group=%G mode=%A atime=%X mtime=%Y ctime=%Z size=%s %N'"'"' /var/www/dataengineeringpodcast 2> /dev/null || stat -f '"'"'user=%Su group=%Sg mode=%Sp atime=%a mtime=%m ctime=%c size=%z %N%SY'"'"' /var/www/dataengineeringpodcast)'
    [pyinfra.api.connectors.ssh] Running command on 172.104.214.53: (pty=False) sh -c '! test -e /var/www/dataengineeringpodcast || (stat -c '"'"'user=%U group=%G mode=%A atime=%X mtime=%Y ctime=%Z size=%s %N'"'"' /var/www/dataengineeringpodcast 2> /dev/null || stat -f '"'"'user=%Su group=%Sg mode=%Sp atime=%a mtime=%m ctime=%c size=%z %N%SY'"'"' /var/www/dataengineeringpodcast)'
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 0
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 0
    [pyinfra.api.facts] Loaded fact directory (path=/var/www/dataengineeringpodcast)
    [pyinfra.api.facts] Getting fact: directory (path=/var/www/dataengineeringpodcast) (ensure_hosts: (Host(172.104.214.53),))
    [pyinfra.api.facts] Getting fact: directory (path=/var/www/dataengineeringpodcast/.git) (ensure_hosts: (Host(172.104.214.53),))
    [pyinfra.api.connectors.ssh] Running command on 66.175.213.219: (pty=False) sh -c '! test -e /var/www/dataengineeringpodcast/.git || (stat -c '"'"'user=%U group=%G mode=%A atime=%X mtime=%Y ctime=%Z size=%s %N'"'"' /var/www/dataengineeringpodcast/.git 2> /dev/null || stat -f '"'"'user=%Su group=%Sg mode=%Sp atime=%a mtime=%m ctime=%c size=%z %N%SY'"'"' /var/www/dataengineeringpodcast/.git)'
    [pyinfra.api.connectors.ssh] Running command on 172.104.214.53: (pty=False) sh -c '! test -e /var/www/dataengineeringpodcast/.git || (stat -c '"'"'user=%U group=%G mode=%A atime=%X mtime=%Y ctime=%Z size=%s %N'"'"' /var/www/dataengineeringpodcast/.git 2> /dev/null || stat -f '"'"'user=%Su group=%Sg mode=%Sp atime=%a mtime=%m ctime=%c size=%z %N%SY'"'"' /var/www/dataengineeringpodcast/.git)'
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 0
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 0
    [pyinfra.api.facts] Loaded fact directory (path=/var/www/dataengineeringpodcast/.git)
    [172.104.214.53] Ready: src/bn_config/deployments/git_test.py

--> Proposed changes:
    Groups: inventory / wordpress_servers
    [66.175.213.219]   Operations: 1   Commands: 2   
    [172.104.214.53]   Operations: 1   Commands: 3   

--> Beginning operation run...
--> Starting operation: Clone Wordpress site repository 
    [pyinfra.api.operations] Starting operation Clone Wordpress site repository on 66.175.213.219
    [pyinfra.api.connectors.ssh] Running command on 66.175.213.219: (pty=None) sh -c 'cd /var/www/podcastinit && git pull'
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 0
    [pyinfra.api.connectors.ssh] Running command on 66.175.213.219: (pty=None) sh -c 'chown -R wordpress /var/www/podcastinit'
    [pyinfra.api.connectors.ssh] Waiting for exit status...
    [pyinfra.api.connectors.ssh] Command exit status: 0
    [66.175.213.219] Success

--> Results:
    Groups: inventory / wordpress_servers
    [66.175.213.219]   Successful: 1   Errors: 0   Commands: 2/2   
    [172.104.214.53]   Successful: 0   Errors: 0   Commands: 0/3 
```

**Comment by Fizzadar** [comment]
I think I've tracked this down to the fact implementation itself assusming the target repository directory exists. Facts are always executed on all hosts, so in this case the `git_branch` fact is tried for both locations on disk, causing one to fail. The issue can also be triggered locally like so:

```
pyinfra @local fact git_branch:/not/a/directory
--> Loading config...
--> Loading inventory...

--> Connecting to hosts...
    [@local] Connected

--> Gathering facts...
    [@local] sh: line 0: cd: /not/a/directory: No such file or directory
    [@local] Error: could not load fact: git_branch (repo=/not/a/directory)
    Loaded fact git_branch (repo=/not/a/directory)

--> Fact data for: git_branch('/not/a/directory',)
{
    "@local": null
}
```

I believe this has gone unnoticed for some time because the `git.repo` operation also checks if the directory exists, hiding thee issue. I've just pushed the change to check which should fix this in https://github.com/Fizzadar/pyinfra/commit/9ba734f991299a46ebe5e0f738eeb54ba55095c6.

**Comment by Fizzadar** [comment]
Just released `1.3.10` with this fix in!
